src/control/server/storage/types.go

```diff
@@ -75,6 +76,47 @@ type (
 
 	// ScmNamespaces is a type alias for []ScmNamespace that implements fmt.Stringer.
 	ScmNamespaces []ScmNamespace
+
+	// NvmeDeviceHealth represents a set of health statistics for a NVMe device.
+	NvmeDeviceHealth struct {
+		Temp            uint32
+		TempWarnTime    uint32
+		TempCritTime    uint32
+		CtrlBusyTime    uint64
+		PowerCycles     uint64
+		PowerOnHours    uint64
+		UnsafeShutdowns uint64
+		MediaErrors     uint64
+		ErrorLogEntries uint64
+		TempWarn        bool
+		AvailSpareWarn  bool
+		ReliabilityWarn bool
+		ReadOnlyWarn    bool
+		VolatileWarn    bool
+		CtrlrPciAddr    string
+	}
+
+	// NvmeNamespace represents an individual NVMe namespace on a device.
+	NvmeNamespace struct {
+		ID           int32
+		Capacity     int32
+		CtrlrPciAddr string
+	}
+
+	// NvmeController represents a NVMe device controller which includes health
+	// and namespace information.
+	NvmeController struct {
+		Model       string
+		Serial      string
+		PciAddr     string
+		FwRev       string
+		SocketID    int32
+		HealthStats *NvmeDeviceHealth
```

Comment:
binding should return format fully compatible with this so no need for coalescing, do we want to use the structure definition from the bindings directly or translate into this?

Reply:
We'll need to translate, because we don't want anything other than the bdev package to depend on the nvme bindings. I briefly considered the idea of making the bindings import the storage package to use these types directly, but I don't think that would really improve anything. As long as we keep the types compatible, it's trivial to do automatic conversion at the network|native|bindings boundaries.